source/agora/consensus/protocol/Nominator.d

```diff
@@ -1032,11 +1032,14 @@ extern(D):
                 auto data = deserializeFull!ConsensusData(candidate[]);
 
                 if (auto msg = this.ledger.validateConsensusData(data))
-                    assert(0, format!"combineCandidates: Invalid consensus data: %s"(
-                        msg));
-
-                log.trace("Combined consensus data: {}", data.prettify);
-                values ~= data;
+                {
```

Comment:
what is the reason for not asserting? If a node considers a value as a candidate at some point, then it should always be valid from that point.

Reply:
The missing validators can change over time so it can become invalid even though it was at some point in time.

Reply:
If you just return false here, while previously returned true for the same candidate, then you will get an assert in the cpp code and agora will crash anyway:
https://github.com/bosagora/agora/issues/1981#issuecomment-823864764

I have a suggestion in mind to tackle this in the right way: 
1. during 'nomination phase' it is okay to change a vote from 'false' to 'true', but impossible to change it from 'true' to 'false'. 
2. during 'balloting phase' it is not okay to change our votes in any ways.

1. During nomination 
If a node receives a candidate with **more** missing validators than the node knows, then he returns 'kInvalidValue'
If a node receives a candidate with **less** missing validators than the node knows, then it returns 'kInvalidValue' AND it will also query the preimages for all nodes in the quorum from the missing preimage(so he can update his missing validator array, and eventually vote with 'kValidValue')
2. During Balloting
if there is any discrepancy between the missing validators we know and between the candidates, we answer with 'kMaybeValid'

the above is just part of many nomination/slashing bugs improvements, and at some point I will solve them(including the corner case you mentioned)
https://github.com/bosagora/agora/issues/2022

Reply:
But until what you mention is implemented then is it not better for combineCandidates to be able to choose another valid `consensusData` rather than asserting. These corner cases are what cause the build to sometimes fail. Are you sure that it will still crash if we choose another candidate from the set we are combining?

Reply:
If it causes test failures, then it is better to remove the assert, but maybe we can also put a link the issue ticket #2022, so at some point we will solve the real issue.

Chosing an other candidate from the set will not not crash, but I think, it can indefinitely delay the consensus run.

Reply:
Perhaps this will not trigger anymore since the wait for preimages was added to many tests where it sometimes caused this. I will drop the commit from this PR and let's see if it hits the assert.